Adding Documentation Website and Deployment

Crash-GHaun opened the pull request:
Hey Team!

I wanted to get this pull request submitted to start the review process, although I recognize there are probably parts of this that will need some additional help or guidance. 

This request is to add the code of our new documentation website to the base repository. The website is built using the Hugo framework and a Docsy theme. It's then compiled and deployed to AppEngine. 

Notable Files and Folders:

- website/  -- This still contains the website content, however the file structure has changed. 
- website/content/en -- This contains the documentation files Hugo uses to generate the static website
- website/layouts -- This contains static html files used by Hugo to generate the static site. 
- website/static -- This contains static files used by Hugo to generate the static site
- website/themes/docsy -- This linked to the Docsy submodule. 

Extremely important files for additional eyes:
- website/app.yaml -- This file contains the new configuration for deploying to AppEngine. This has been tested but would like a second set of eyes
- .github/workflows/push-tags.yml -- This file is critical for this PR and needs an extra set of eyes and I could use some help ensuring this code works. I have **not** tested this as I wasn't sure of the correct process to do so. We must not submit this pull request until this has been tested and confirmed to work. 

I think this sums it up but I'm sure I'll add more information here :) 

Thanks!

simonz130 commented on website/content/en/_index.html:
```diff
@@ -0,0 +1,81 @@
+	<a class="btn btn-lg btn-secondary mr-3 mb-4" href="https://github.com/google/docsy-example">
+		Download <i class="fab fa-github ml-2 "></i>
+	</a>
+	<p class="lead mt-5">Porridge temperature assessment - in the cloud!</p>
```
What is this text about?

Crash-GHaun replied:
Actually this is an unused file, I'll delete it!

Crash-GHaun commented on .gitmodules:
```diff
@@ -0,0 +1,3 @@
+[submodule "website/themes/docsy"]
+	path = website/themes/docsy
+	url = https://github.com/google/docsy.git
```
When you say consume it as a package do you mean uploading the files directly in the repo? We absolutely can do that, although this keeps our repo cleaner in my opinion. Open to other solutions :)

Daniel-Sanche replied:
My vote would probably be to containerize things, and pull docsy in as a dependency when building the dockerfile. But I'm not sure if there was a reason to use appengine standard

simonz130 replied:
Given that we own cloud-ops-sandbox.dev domain now, we can host the website wherever we want. So we can move it to firebase, keep it on AppEngine Standard, move to Flex - whatever makes sense for hosting static website and handles SSL for us "for free".

Crash-GHaun replied:
Just talked with Daniel, and I would like to +1 to containerizing. Unless you think otherwise that's probably out of scope for this PR but is something that should be considered for the near future. Lmk your thoughts

Crash-GHaun commented on website/README.md:
```diff
@@ -4,23 +4,24 @@ The Cloud Operations Sandbox website is currently deployed at [cloud-ops-sandbox
-* `main.css` - located in `website/css/` directory. This is the style sheet for Stackdriver Sandbox's website.
-* `images/` - located in `website/` directory. Any images on the website should be located in this directory.
+* `config.toml` - Located in `website/`, this file specifies the hugo properties for the website. This is critical for hugo to build the website
+* `deploy.sh` - Not sure I need this actually
```
yup, I'll do some testing to confirm we can remove deploy.sh :)

Crash-GHaun commented on website/config.toml:
```diff
@@ -0,0 +1,162 @@
+[services]
+[services.googleAnalytics]
+# Comment out the next line to disable GA tracking. Also disables the feature described in [params.ui.feedback].
+id = "UA-137862890-1"
```
I did pull this ID from the original site. I'll double check to confirm in case something reverted.

Crash-GHaun commented on website/config.toml:
```diff
@@ -0,0 +1,162 @@
+theme = ["docsy"]
+
+# Will give values to .Lastmod etc.
+enableGitInfo = true
```
This enabled the website to have Hugo add Git information like "Last Modified" to the site
https://gohugo.io/variables/git/

Daniel-Sanche commented on website/config.toml:
```diff
@@ -0,0 +1,162 @@
+
+enableRobotsTXT = true
+
+theme = ["docsy"]
```
nit: maybe it's me, but I'm not a huge fan of the table of contents style of this theme. I find it hard to navigate

simonz130 replied:
Any suggestions on which theme would be more useful?

Crash-GHaun replied:
Open to suggestions! The open source team recommended Docsy but since this is a Theme it should be easy to change. More than happy to roll out with this and change it in the future if we have a theme we prefer.

Daniel-Sanche replied:
Sorry I'm not familiar with the theme options. Docsy is probably fine since I've seen it in other Google products. I'm just not a fan of those nav bars

Daniel-Sanche commented on website/content/en/docs/getting_started.md:
```diff
@@ -0,0 +1,53 @@
+  * [Set Up](#set-up)
+  * [Next Steps](#next-steps)
+  * [Clean Up](#clean-up)
+* [Service Overview](/docs/service_overview)
```
This link seems to be broken

Crash-GHaun replied:
Service Overview seems to work for me. Can you confirm it's still broken for you?

Daniel-Sanche commented on website/content/en/docs/getting_started.md:
```diff
@@ -0,0 +1,53 @@
+weight: 2
+---
+{{% pageinfo %}}
+* Using Sandbox
```
should this be a link?

Daniel-Sanche commented:
It looks like the license check CI is failing. Every source file should have an open source header. You can use this tool to add them all: https://github.com/google/addlicense

Daniel-Sanche commented:
Since this website is starting to get more complicated, it might be useful to add a CI test around it, and/or add a way to spin it up quickly (either on GAE or locally) to the Makefile.

This might be better as an issue for the future, but let me know what you think

minherz commented:
> It looks like the license check CI is failing. Every source file should have an open source header. You can use this tool to add them all: https://github.com/google/addlicense

Shamelessly PR`ing my extension for vscode -- [Copyright Inserter](https://marketplace.visualstudio.com/items?itemName=minherz.copyright-inserter) 😃

simonz130 commented on website/config.toml:
```diff
@@ -0,0 +1,162 @@
+section = ["HTML", "print"]
+
+[params]
+copyright = "Cloud Sandbox"
```
It should say "Cloud Operations Sandbox authors".

simonz130 commented on website/config.toml:
```diff
@@ -0,0 +1,162 @@
+# The version number for the version of the docs represented in this doc set.
+# Used in the "version-banner" partial to display a version number for the 
+# current doc set.
+version = "1.0"
```
Can we have the release version show up here, or the version of the build (commit sha)?
@Daniel-Sanche for additional thoughts.

Crash-GHaun replied:
I think we can do this in push-tags.yml. @Daniel-Sanche  to confirm

Daniel-Sanche replied:
We could have [`make_release.sh`](https://github.com/GoogleCloudPlatform/cloud-ops-sandbox/blob/master/make-release.sh) come through here and replace the version number on each release if we want. That's how version numbers are set in other places in the codebase

Although the more files it touches, the more I feel like we should consider other options too though

Daniel-Sanche commented on .github/workflows/push-tags.yml:
```diff
@@ -43,8 +43,17 @@ jobs:
+      run: sudo apt-get install hugo npm
+    - name: Install PostCSS
+      timeout-minutes: 20
+      run: sudo npm install -g autoprefixer postcss-cli postcss
```
I added these dependencies to the GitHub Actions worker VM. Could you add them to the [dependencies script there](https://github.com/GoogleCloudPlatform/cloud-ops-sandbox/blob/master/.github/workflows/install-dependencies.sh)?

```
# install nodejs/hugo
curl -fsSL https://deb.nodesource.com/setup_lts.x | sudo -E bash -
sudo apt-get install -y nodejs hugo
sudo npm install -g autoprefixer postcss-cli postcs
```

Crash-GHaun replied:
On it now, will get it done this afternoon

Daniel-Sanche commented on .github/workflows/push-tags.yml:
```diff
@@ -43,8 +43,17 @@ jobs:
+      run: sudo npm install -g autoprefixer postcss-cli postcss
+    - name: Generate Website
+      timeout-minutes: 20
+      run: hugo --source=website/ --destination=deploy/
```
running this seemed caused a weird behavior error on the VM. It looks like it's creating infinite nested deploy folders:

```
Error: Error copying static files to /home/sanche/cloud-ops-sandbox/website/deploy: stat /home/sanche/cloud-ops-sandbox/website/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/deploy/content/en/docs/User Guide/Learn Ops Management/cloud_error_reporting.md: file name too long
```

daniel-sanche replied:
I'm having trouble getting this command to run. Maybe we should just commit all these files to the repo for now, and transition to docker containers later?

Crash-GHaun replied:
Hey! I can throw some time on your calendar to walk through it and see what is going on if you prefer. It does look like that's a bug of some kind.

Crash-GHaun replied:
Website content has been generated and added to the PR

daniel-sanche commented:
Just wanted to summarize what I see as the existing issues. If we want to get this in the next release, we should be able to open issues to address most of these later:

- [ ] [We should make a containerization plan](https://github.com/GoogleCloudPlatform/cloud-ops-sandbox/pull/641/files/a12341b2c45628bc8705d36779120dee3e27f81f#r593485312)
- [ ] [I'm having trouble getting hugo to build](https://github.com/GoogleCloudPlatform/cloud-ops-sandbox/pull/641/files/a12341b2c45628bc8705d36779120dee3e27f81f#r604463290). Maybe commit files directly?
- [ ] [If we keep the hugo-compilation step in the CI, we'll need to add some steps to the install script](https://github.com/GoogleCloudPlatform/cloud-ops-sandbox/pull/641/files/a12341b2c45628bc8705d36779120dee3e27f81f#r601874333)
- [ ] [should we edit the version number?](https://github.com/GoogleCloudPlatform/cloud-ops-sandbox/pull/641/files/a12341b2c45628bc8705d36779120dee3e27f81f#r599840625)
- [ ] [I thought I saw a broken link, but haven't been able to reproduce](https://github.com/GoogleCloudPlatform/cloud-ops-sandbox/pull/641/files/a12341b2c45628bc8705d36779120dee3e27f81f#r599138314)

Crash-GHaun commented:
Created issue https://github.com/GoogleCloudPlatform/cloud-ops-sandbox/issues/662 for website containerization. 

Generated the website locally and added to the PR/commit.

 Don't need to have the hugo dependencies now as we aren't generating via github actions
 
 Added issue: https://github.com/GoogleCloudPlatform/cloud-ops-sandbox/issues/663 to move toward updating the website with a version number.

daniel-sanche commented:
Great, thanks for making these changes! As long as the tests pass, this should be good to merge
